File: dart/transaction_system/dart_application_1/bin/transaction.dart
// lib/models/transaction.dart

enum TransactionType { stockIn, stockOut, adjustment, transfer }

class Transaction {
  final String id;
  final DateTime timestamp;
  final TransactionType type;
  final String itemId;
  final String itemSku;
  final int quantityDelta; // positive = added, negative = removed
  final int quantityBefore;
  final int quantityAfter;
  final double unitCost;
  final String? referenceId;  // e.g. PO number, invoice number
  final String? performedBy;
  final String? notes;

  const Transaction({
    required this.id,
    required this.timestamp,
    required this.type,
    required this.itemId,
    required this.itemSku,
    required this.quantityDelta,
    required this.quantityBefore,
    required this.quantityAfter,
    required this.unitCost,
    this.referenceId,
    this.performedBy,
    this.notes,
  });

  double get totalValue => quantityDelta.abs() * unitCost;

  Map<String, dynamic> toJson() => {
        'id': id,
        'timestamp': timestamp.toIso8601String(),
        'type': type.name,
        'item_id': itemId,
        'item_sku': itemSku,
        'quantity_delta': quantityDelta,
        'quantity_before': quantityBefore,
        'quantity_after': quantityAfter,
        'unit_cost': unitCost,
        'reference_id': referenceId,
        'performed_by': performedBy,
        'notes': notes,
      };

  factory Transaction.fromJson(Map<String, dynamic> json) => Transaction(
        id: json['id'] as String,
        timestamp: DateTime.parse(json['timestamp'] as String),
        type: TransactionType.values.byName(json['type'] as String),
        itemId: json['item_id'] as String,
        itemSku: json['item_sku'] as String,
        quantityDelta: json['quantity_delta'] as int,
        quantityBefore: json['quantity_before'] as int,
        quantityAfter: json['quantity_after'] as int,
        unitCost: (json['unit_cost'] as num).toDouble(),
        referenceId: json['reference_id'] as String?,
        performedBy: json['performed_by'] as String?,
        notes: json['notes'] as String?,
      );

  @override
  String toString() =>
      'Transaction(id: $id, type: ${type.name}, sku: $itemSku, delta: $quantityDelta)';
}

File: python system/inventory_system/models/transaction.py
from enum import Enum
from datetime import datetime
from typing import Optional, Dict, Any


class TransactionType(Enum):
    STOCK_IN = "stockIn"
    STOCK_OUT = "stockOut"
    ADJUSTMENT = "adjustment"
    TRANSFER = "transfer"


class Transaction:
    def __init__(
        self,
        id: str,
        timestamp: datetime,
        type: TransactionType,
        item_id: str,
        item_sku: str,
        quantity_delta: int,
        quantity_before: int,
        quantity_after: int,
        unit_cost: float,
        reference_id: Optional[str] = None,
        performed_by: Optional[str] = None,
        notes: Optional[str] = None,
    ):
        self.id = id
        self.timestamp = timestamp
        self.type = type
        self.item_id = item_id
        self.item_sku = item_sku
        self.quantity_delta = quantity_delta
        self.quantity_before = quantity_before
        self.quantity_after = quantity_after
        self.unit_cost = unit_cost
        self.reference_id = reference_id
        self.performed_by = performed_by
        self.notes = notes

    @property
    def total_value(self) -> float:
        return abs(self.quantity_delta) * self.unit_cost

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": self.id,
            "timestamp": self.timestamp.isoformat(),
            "type": self.type.value,
            "item_id": self.item_id,
            "item_sku": self.item_sku,
            "quantity_delta": self.quantity_delta,
            "quantity_before": self.quantity_before,
            "quantity_after": self.quantity_after,
            "unit_cost": self.unit_cost,
            "reference_id": self.reference_id,
            "performed_by": self.performed_by,
            "notes": self.notes,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "Transaction":
        return cls(
            id=data["id"],
            timestamp=datetime.fromisoformat(data["timestamp"]),
            type=TransactionType(data["type"]),
            item_id=data["item_id"],
            item_sku=data["item_sku"],
            quantity_delta=data["quantity_delta"],
            quantity_before=data["quantity_before"],
            quantity_after=data["quantity_after"],
            unit_cost=float(data["unit_cost"]),
            reference_id=data.get("reference_id"),
            performed_by=data.get("performed_by"),
            notes=data.get("notes"),
        )

    def __str__(self) -> str:
        return f"Transaction(id={self.id}, type={self.type.value}, sku={self.item_sku}, delta={self.quantity_delta})"